doc/tools/coqrst/coqdomain.py

```diff
@@ -400,57 +400,101 @@ class TableObject(NotationObject):
+    i.e. displayed in 3 columns, however
+    unlike ``.. productionlist``\ s, this directive accepts notation syntax.
 
     Example::
 
-        .. prodn:: term += let: @pattern := @term in @term
         .. prodn:: occ_switch ::= { {? {| + | - } } {* @num } }
+        .. prodn:: term += let: @pattern := @term in @term
+        .. prodn:: | second_production
+
+       The first line defines "occ_switch", which must be unique in the document.  The second
+       references but doesn't define "term".  The third form is for continuing the
+       definition of a nonterminal when it has multiple productions.  It leaves the first
+       column in the output blank.
 
     """
     subdomain = "prodn"
     #annotation = "Grammar production"
 
+    prev_position = None
+    table = None
+    tbody = None
+
     def _render_signature(self, signature, signode):
         raise NotImplementedError(self)
 
     SIG_ERROR = ("{}: Invalid syntax in ``.. prodn::`` directive"
                  + "\nExpected ``name ::= ...`` or ``name += ...``"
-                 + " (e.g. ``pattern += constr:(@ident)``)")
+                 + " (e.g. ``pattern += constr:(@ident)``)\n"
+                 + "  in `{}`")
 
     def handle_signature(self, signature, signode):
-        nsplits = 2
-        parts = signature.split(maxsplit=nsplits)
-        if len(parts) != 3:
-            loc = os.path.basename(get_node_location(signode))
-            raise ExtensionError(ProductionObject.SIG_ERROR.format(loc))
-
-        lhs, op, rhs = (part.strip() for part in parts)
-        if op not in ["::=", "+="]:
-            loc = os.path.basename(get_node_location(signode))
-            raise ExtensionError(ProductionObject.SIG_ERROR.format(loc))
-
-        self._render_annotation(signode)
-
-        lhs_op = '{} {} '.format(lhs, op)
-        lhs_node = nodes.literal(lhs_op, lhs_op)
+        parts2 = signature.split(maxsplit=1)
+        if parts2[0].strip() == "|":
+            lhs = ""
+            op = "|"
+            rhs = parts2[1].strip()
+        else:
+            nsplits = 2
+            parts = signature.split(maxsplit=nsplits)
+            if len(parts) != 3:
+                loc = os.path.basename(get_node_location(signode))
+                raise ExtensionError(ProductionObject.SIG_ERROR.format(loc, signature))
+            else:
+                lhs, op, rhs = (part.strip() for part in parts)
+                if op not in ["::=", "+="]:
+                    loc = os.path.basename(get_node_location(signode))
+                    raise ExtensionError(ProductionObject.SIG_ERROR.format(loc, signature))
 
         position = self.state_machine.get_source_and_line(self.lineno)
+        if lhs == '':
+            lhs_node = nodes.literal('', '')
+        else:
+            lhs_node = notation_to_sphinx('@'+lhs, *position)
         rhs_node = notation_to_sphinx(rhs, *position)
-        signode += addnodes.desc_name(signature, '', lhs_node, rhs_node)
+
+        # create a row
```

Comment:
This table-based code looks reasonable to be, but I wonder if you would have had better luck with formatting on the LaTeX side by keeping the `addnodes.desc_name` nodes and adding custom classes

Reply:
(And you could have used CSS on the HTML side to make it look the way you want it)

Reply:
I experimented with the signature code for about a week.  It generated invalid .tex files (apparently expecting 3 nested signature nodes).  It also generated the Permalink, so I needed to include the <table> unadorned by signature nodes.

A had to do a lot of research and experimentation along the way to get to this point.  I have no experience with latex.  How would I apply custom classes to latex?

> (And you could have used CSS on the HTML side to make it look the way you want it)

How do you line up columns in a table-like format without <table>?  Or for that matter in latex?

Reply:
> How would I apply custom classes to latex?

That's what the `\DURole` command does, and I agree with you it isn't well doumented.  `DURole` is for "docutils role"; see here: https://docutils.readthedocs.io/en/sphinx-docs/user/latex.html#custom-interpreted-text-roles

> How do you line up columns in a table-like format without 

You could use a table on both sides, but you would be able to customize the look of it.  
Alternatively, you can do table-like layout in CSS; see https://stackoverflow.com/questions/4309950/how-to-align-input-forms-in-html/23741073#23741073